pkg/kn/commands/service/create.go

```diff
@@ -260,3 +267,32 @@ func constructService(cmd *cobra.Command, editFlags ConfigurationEditFlags, name
 	}
 	return &service, nil
 }
+
+// constructServiceFromFile creates struct from provided file
+func constructServiceFromFile(cmd *cobra.Command, editFlags ConfigurationEditFlags, name, namespace string) (*servingv1.Service, error) {
+	var service servingv1.Service
+	file, err := os.Open(editFlags.File)
+	if err != nil {
+		return nil, err
+	}
+	decoder := yaml.NewYAMLOrJSONDecoder(file, 512)
+
+	err = decoder.Decode(&service)
+	if err != nil {
+		return nil, err
+	}
+	if service.Name != name {
+		return nil, fmt.Errorf("provided service name '%s' doesn't match name from file '%s'", name, service.Name)
+	}
+	// Set namespace in case it's specified as --namespace
+	service.ObjectMeta.Namespace = namespace
+
+	// We need to generate revision to have --force replace working
+	revName, err := servinglib.GenerateRevisionName(editFlags.RevisionName, &service)
```

Comment:
I am thinking if the client-side revision name generation could be turned off if using files mode?

Reply:
```
➜  client git:(create-from-file) ✗ cat /tmp/hello.yaml
apiVersion: serving.knative.dev/v1
kind: Service
metadata:
  name: helloworld-go
spec:
  template:
    spec:
      containers:
        - image: gcr.io/knative-samples/helloworld-go
          env:
            - name: TARGET
              value: "Go Sample v1"
➜  client git:(create-from-file) ✗ kn service create helloworld-go -f /tmp/hello.yaml 
Creating service 'helloworld-go' in namespace 'default':
  0.028s The Route is still working to reflect the latest desired specification.
  0.061s Configuration "helloworld-go" is waiting for a Revision to become ready.
  0.106s ...
  3.751s ...
  3.795s Ingress has not yet been reconciled.
  3.868s Waiting for load balancer to be ready
  4.190s Ready to serve.
Service 'helloworld-go' created to latest revision 'helloworld-go-5j5mp' is available at URL:
http://helloworld-go.default.example.com
➜  client git:(create-from-file) ✗ kn service create helloworld-go -f /tmp/hello.yaml --force
Replacing service 'helloworld-go' in namespace 'default':
```

When I was testing the flag locally I've experienced the above behaviour. When trying to -recreate the service with `--force` the update hangs and timeouts eventually. However it's not happening for `kn service update` flow or imperative `kn service create` due to the fact that we generate revision name by default. May it a serving bug?

Reply:
yes could be, but we should be sure what it actually happen when using `--force`. Could you do a debug session and check where exactly it hangs ?

For opening an issue for serving we would need to recreate the steps via yaml files, I think.

Reply:
Well, the issue is limited to a corner case when the same source is used twice, in other words to replace the same already created same Service. In that case such an update op never replaces the original Service content, causing no set of event to be produced in the cluster for wait to finish without timing out.
There's only one synthetic event `ADDED` [received](https://github.com/knative/client/blob/master/pkg/wait/wait_for_ready.go#L179), but nothing else as the server-side doesn't update anything at all.
The same behaviour can be reproduced with `kubectl replace`, meaning that original service isn't actually replaced, is still the same with original timestamp etc., but there's no wait for ready that could be blocking the execution.

That kind of issue isn't present in imperative `create` flow due to the fact that [generated revision name](https://github.com/knative/client/blob/master/pkg/kn/commands/service/configuration_edit_flags.go#L334-L341) is enough of a change to trigger the update/replace mechanism, therefore desired set of events is received in wait function.


I'd suggest the following:

- Implement @rhuss's suggestion from [this comment  ](https://github.com/knative/client/pull/913#discussion_r450685124)
- Mark `filename` flag with `markFlagMakesRevision` to enforce revision generation
  - Adding modification from other flag would always trigger new revision name generation, therefore we'll benefit from it for filename as well (although it might be seen as hack as well)
- => Profit.

Reply:
For a quick win, I would go for the second option. We then can implement that improvement to pick up other options later. wdyt ?

Reply:
OK, I'm on board with quick win. :) Anyway it'll better to have improvement PR with proper amount of tests and bake time for options combination. I'll create a new issue for it.

Reply:
Follow-up enhancement issue https://github.com/knative/client/issues/923.